lib/loader/index.js

```diff
@@ -214,6 +301,117 @@ function postInstantiate(baseModule, instance) {
 
   baseModule.__getArray = __getArray;
 
+  function __getUint8Array(arr) {
+    // check memory
+    const buffer = memory.buffer;
+    if (U32.buffer !== buffer) {
+      U8  = new Uint8Array(buffer);
+      U32 = new Uint32Array(buffer);
+    }
+    const buf = U32[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
+    const length = U32[buf + SIZE_OFFSET >>> 2];
+    return U8.slice(buf, buf + length);
```

Comment:
> This line seems to be the difference between all the different functions
Nope. Also [checking memory](https://github.com/AssemblyScript/assemblyscript/pull/794/files/f66c2be64df30c2c197b2bab47eb432e93cc6c6c#diff-d7e2f1f33f6cad0b0feae844c752f106R322).

Reply:
It can't be [`null`](https://github.com/AssemblyScript/assemblyscript/blob/5e918d868753a716d7de0b019259c4da7722f7b8/lib/loader/index.js#L116)

Reply:
You mean if `buf` is `null`?

Reply:
Might even be that a `new Uint32Array(someBuffer)` has zero cost in modern VMs and caching `U32` isn't actually necessary. Haven't tested this yet, but always wanted to. If the case, that might help to "generate" these functions, instead of implementing 10 different ones.

Reply:
`wasm-bindgen` use this util:
```js
let cachegetInt32Memory = null;
function getInt32Memory() {
    if (cachegetInt32Memory === null || cachegetInt32Memory.buffer !== wasm.memory.buffer) {
        cachegetInt32Memory = new Int32Array(wasm.memory.buffer);
    }
    return cachegetInt32Memory;
}
```

Reply:
So I'll test this approach. See how this will affect to performance

Reply:
Question is: If there's some `buffer`, what does a modern VM generate on `new Uint32Array(buffer, start, end)[offset]`? Does this produce an actual allocation through `new`, or does this simply optimize down to a simple 32-bit integer load? In the latter case, the whole U32 etc. stuff could be unnecessary.

Reply:
Baseline (PR) perf result: `589,885 ops/sec`

1) Approach suggested by @willemneal:
```js
function makeArray(TypedCtor){
  var arr = new TypedCtor(memory.buffer);
  return () => {
    var buf = memory.buffer;
    if (arr.buffer !== buf) { arr = new TypedCtor(buf); }
    return arr;
  }
}

const U32 = makeArray(Uint32Array);
const U8  = makeArray(Uint8Array);

function __getUint8Array(arr) {
  const buf = U32()[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
  const length = U32()[buf + SIZE_OFFSET >>> 2];
  return U8().slice(buf, buf + length);
}
```
Result: `529,102 ops/sec`

2) Rust's style:
```js
  function getU32() {
    if (cachedU32 === null || cachedU32.buffer !== memory.buffer) {
      cachedU32 = new Uint32Array(memory.buffer);
    }
    return cachedU32;
  }

  function getU8() {
    if (cachedU8 === null || cachedU8.buffer !== memory.buffer) {
      cachedU8 = new Uint8Array(memory.buffer);
    }
    return cachedU8;
  }
function __getUint8Array(arr) {
    const buf = getU32()[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
    const length = getU32()[buf + SIZE_OFFSET >>> 2];
    return getU8().slice(buf, buf + length);
  }
```
Result: `538,241 ops/sec`

3) Same but with caching `getU32()`:
```js
const u32 = getU32();
const buf = u32[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
const length = u32[buf + SIZE_OFFSET >>> 2];
return getU8().slice(buf, buf + length);
```
Result: `562,116 ops/sec`

Reply:
I think current approach archive best possible performance with acceptable code size increasing.

Reply:
For comparison purposes, how does no caching at all perform?

```ts
function __getInt8Array(arr) {
  const u32 = new Uint32Array(memory.buffer);
  const buf = u32[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
  const length = u32[buf + SIZE_OFFSET >>> 2];
  return new Uint8Array(memory.buffer).slice(buf, buf + length);
}
```

Reply:
@dcodeIO 
Result: `548,346 ops/sec`. Similar to 2).

Reply:
Hmm, that's interesting. Shouldn't that be significantly slower? Seems to somewhat indicate that caching isn't giving us much.

Reply:
```js
function __getUint8Array(arr) {
  const u32 = new Uint32Array(memory.buffer);
  const buf = u32[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
  const length = u32[buf + SIZE_OFFSET >>> 2];
  return new Uint8Array(memory.buffer).slice(buf, buf + length);
}
```

Result: `577,601 ops/sec`

Reply:
So, what about something like

```ts
function haveTheCakeAndEatItToo(theArray) {
  return function(arr) {
    const u32 = new Uint32Array(memory.buffer);
    const buf = u32[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
    const length = u32[buf + SIZE_OFFSET >>> 2];
    return new theArray(memory.buffer).slice(buf, buf + length);
  }
}
var __getUint8Array = haveTheCakeAndEatItToo(Uint8Array);
```

Reply:
```js
function __getUint8Array(arr) {
  var buffer = memory.buffer;
  var u32 = new Uint32Array(buffer);
  var buf = u32[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
  var length = u32[buf + SIZE_OFFSET >>> 2];
  return new Uint8Array(buffer).slice(buf, buf + length);
}
```

Result: `592,759 ops/sec`

I guess it's optimal!

Reply:
> the closure will make it slightly slower

My expectation would be that it doesn't, but as you said, even if it does, maybe a `.bind` doesn't.

Reply:
Currently I [reduce footprint](https://github.com/AssemblyScript/assemblyscript/pull/794/commits/860da936df6a6196d449e676474c14088922d493). This should be optimal for speed and size

Reply:
Now, what if

```ts
function getTypedArray(Type, shift, arr) {
  var buffer = memory.buffer;
  var u32 = new Uint32Array(buffer);
  var buf = u32[arr + ARRAYBUFFERVIEW_DATASTART_OFFSET >>> 2];
  var length = u32[buf + SIZE_OFFSET >>> 2];
  return new Type(buffer).slice(buf >>> shift, buf + length >>> shift);
}

baseModule.__getUint8Array = getTypedArray.bind(null, Uint8Array, 1);
...
```

Reply:
Yeah. Also great!

Reply:
Done!

Reply:
@jtenner `597,915 ops/sec`